### PR TITLE
Emit all sections as before except for debug_line and debug_frames which are delayed

### PR DESCRIPTION
**Pull request by poechsel**
#1643 caused problem with mdx as it's involved a big changed in signatures.

This PR does the same but explicitly. All sections - but .debug_line, and later .debug_frames are emitted directly. .debug_line and .debug_frames are now considered as "delayed" and will only be emitted once [Dwarf.emit_delayed] is called.

### CODE REVIEW

**Comment by mshinwell** [comment]
I'm reviewing this

File: backend/asm_targets/asm_section.mli [hunk 1]
```diff
@@ -54,6 +54,10 @@ type section_details = private
 
 val dwarf_sections_in_order : unit -> t list
 
+(* If [is_delayed] = true the corresponding function will be emitted after
```
Comment (mshinwell) [inline backend/asm_targets/asm_section.mli:57]: This should say "section"

File: backend/internal_assembler/internal_assembler.ml [hunk 2]
```diff
@@ -170,28 +170,30 @@ let assemble_one_section ~name instructions =
...
+          List.fold_left
+            (fun acc i ->
+              match i with X86_ast.Align (data, n) when n > acc -> n | _ -> acc)
+            0 instructions
+        in
+        Section_name.Map.add name
+          ( align,
+            X86_binary_emitter.assemble_section X64
+              { X86_binary_emitter.sec_name = X86_proc.Section_name.to_string name;
+                sec_instrs = Array.of_list instructions
```
Comment (mshinwell) [inline backend/internal_assembler/internal_assembler.ml:188]: The useful comment that was here appears to have been lost?

File: backend/debug/dwarf/dwarf_ocaml/dwarf.ml [hunk 3]
```diff
@@ -103,3 +109,18 @@ let emit t ~basic_block_sections ~binary_backend_available =
   Profile.record "emit_dwarf"
     (emit ~basic_block_sections ~binary_backend_available)
     t
+
+let emit_delayed t ~basic_block_sections ~binary_backend_available =
+  if t.emitted_delayed
+  then
+    Misc.fatal_error
+      "Cannot call [Dwarf.emit] more than once on a given value of type \
```
Comment (mshinwell) [inline backend/debug/dwarf/dwarf_ocaml/dwarf.ml:117]: This should say Dwarf.emit_delayed

File: backend/x86_proc.ml [hunk 4]
```diff
@@ -338,20 +339,22 @@ let directive dir =
...
-        Section_name.Tbl.add asm_code_by_section name !asm_code_current_section)
+        Section_name.Tbl.add where name !asm_code_current_section)
   | dir -> !asm_code_current_section := dir :: !(!asm_code_current_section)
 
 let emit ins = directive (Ins ins)
 
 let reset_asm_code () =
   asm_code := [];
   asm_code_current_section := ref [];
+  (* Section_name.Tbl.clear delayed_sections; *)
```
Comment (mshinwell) [inline backend/x86_proc.ml:357]: Is this needed?  If not let's delete the line.

**Review by mshinwell — CHANGES_REQUESTED** [review]

**Review by mshinwell — APPROVED** [review]

**Review by xclerc — APPROVED** [review]

**Pull request merged**